File: ios/objective-c/v2/AmpliObjectiveCSampleApp/AmpliObjectiveCSampleAppTests/AmpliTests.m
//
//  AmpliTests.m
//  AmpliObjectiveCSampleAppTests
//
//  Created by Qingzhuo Zhen on 12/2/21.
//

#import <XCTest/XCTest.h>
@import AmplitudeSwift;
#import "Ampli.h"

@interface AmpliTests : XCTestCase
@property (nonatomic, strong) Ampli *ampli;
@property (nonatomic, weak) XCTestExpectation* pluginRun;
@end

@implementation AmpliTests

- (void)setUp {
    _ampli = [Ampli new];
    _pluginRun = [self expectationWithDescription:@"Wait for plugin run"];
}

- (void) initAmpliWithNewInstance:(NSString *) instanceName {
    AMPConfiguration* configuration = [AMPConfiguration initWithApiKey:@"test-api-key" instanceName:instanceName];
    configuration.defaultTracking = AMPDefaultTrackingOptions.NONE;
    configuration.migrateLegacyData = false;
    LoadClientOptions* clientOptions = [LoadClientOptions initWithConfiguration:configuration];
    [_ampli load:[LoadOptions initWithClientOptions:clientOptions]];
}

- (void)testTrackWithNoProperies {
    [self initAmpliWithNewInstance:@"testTrackWithNoProperties"];

    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        XCTAssertEqualObjects(event.eventType, @"Event No Properties");
        [self->_pluginRun fulfill];
        return event;
    }]];

    [_ampli track:[EventNoProperties build]];
    [_ampli flush];

    [self waitForExpectationsWithTimeout:1.0 handler:^(NSError *error) {
        if (error) {
            XCTFail(@"test timed out");
        }
    }];
}

- (void)testTrackEventWithAllTypes {
    [self initAmpliWithNewInstance:@"testTrackEventWithAllTypes"];

    NSMutableDictionary *extraDict = [NSMutableDictionary new];
    [extraDict setObject:@"extra test" forKey:@"test"];
    NSArray *requiredArray = [NSArray arrayWithObjects:@"array element 1", @"array element 2", nil];

    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        AMPProperties* eventProperties = event.eventProperties;
        XCTAssertEqualObjects([eventProperties get:@"requiredArray"], requiredArray);
        XCTAssertEqualObjects([eventProperties get:@"requiredBoolean"], @true);
        XCTAssertEqualObjects([eventProperties get:@"requiredEnum"], @"Enum1");
        XCTAssertEqualObjects([eventProperties get:@"requiredInteger"], @10);
        XCTAssertEqualObjects([eventProperties get:@"requiredNumber"], @2.0F);
        XCTAssertEqualObjects([eventProperties get:@"requiredString"], @"required string");
        XCTAssertNil([eventProperties get:@"optionalString"]);
        [self->_pluginRun fulfill];
        return event;
    }]];

    [_ampli track:[EventWithAllProperties requiredArray:requiredArray
                                         requiredBoolean:true
                                         requiredEnum:EventWithAllPropertiesRequiredEnumEnum1
                                         requiredInteger:10
                                         requiredNumber:2.0F
                                         requiredString:@"required string"
    ]];
    [_ampli flush];

    [self waitForExpectationsWithTimeout:2.0 handler:^(NSError *error) {
        if (error) {
            NSLog(@"timeout errored: %@", error);
            XCTFail(@"test timed out");
        }
    }];
}

- (void)testIdentify {
    [self initAmpliWithNewInstance:@"testIdentify"];

    NSString* userId = @"test-user-id";
    NSString* deviceId = @"test-device-id";
    AMPEventOptions* eventOptions = [AMPEventOptions new];
    eventOptions.deviceId = deviceId;
    eventOptions.userId = userId;
    
    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        XCTAssertEqualObjects(event.eventType, @"$identify");
        XCTAssertEqualObjects(event.userId, userId);
        XCTAssertEqualObjects(event.deviceId, deviceId);
        [self->_pluginRun fulfill];
        return event;
    }]];
    
    [_ampli identify:userId
           identify:[Identify requiredNumber: 22.0F builderBlock:^(IdentifyBuilder *b) {
                b.optionalArray = [NSArray arrayWithObjects:@"optional string", nil];
           }]
           options:eventOptions
   ];
    [_ampli flush];
    [self waitForExpectationsWithTimeout:2.0 handler:^(NSError *error) {
        if (error) {
            NSLog(@"timeout errored: %@", error);
            XCTFail(@"test timed out");
        }
    }];
}

- (void)testIdentifyUserIdOnEvent {
    [self initAmpliWithNewInstance:@"testIdentifyUserIdOnEvent"];

    NSString *eventOptionsUserId = @"test-user-id-options";
    NSString *deviceId = @"test-device-id";
    AMPEventOptions* eventOptions = [AMPEventOptions new];
    eventOptions.deviceId = deviceId;
    eventOptions.userId = eventOptionsUserId;

    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        XCTAssertEqualObjects(event.eventType, @"$identify");
        XCTAssertEqualObjects(event.userId, eventOptionsUserId);
        XCTAssertEqualObjects(event.deviceId, deviceId);
        XCTAssertEqualObjects([self->_ampli.client getUserId], eventOptionsUserId);
        [self->_pluginRun fulfill];
        return event;
    }]];
    
    [_ampli identify:nil
           identify:[Identify requiredNumber: 22.0F builderBlock:^(IdentifyBuilder *b) {
                b.optionalArray = [NSArray arrayWithObjects:@"optional string", nil];
           }]
           options:eventOptions
   ];
    [_ampli flush];

    [self waitForExpectationsWithTimeout:2.0 handler:^(NSError *error) {
        if (error) {
            NSLog(@"timeout errored: %@", error);
            XCTFail(@"test timed out");
        }
    }];
}

- (void)testSetGroup {
    [self initAmpliWithNewInstance:@"testSetGroup"];

    NSString *groupType = @"test group type";
    NSString *groupName = @"test group name";

    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        XCTAssertEqualObjects(event.eventType, @"$identify");
        AMPProperties* userProperties = event.userProperties;
        NSMutableDictionary *userPropertiesSet = [userProperties get:@"$set"];
        XCTAssertEqualObjects(userPropertiesSet[groupType], groupName);
        [self->_pluginRun fulfill];
        return event;
    }]];
    
    [_ampli.client setGroup:groupType groupName:groupName];
    [_ampli flush];

    [self waitForExpectationsWithTimeout:2.0 handler:^(NSError *error) {
        if (error) {
            NSLog(@"timeout errored: %@", error);
            XCTFail(@"test timed out");
        }
    }];
}

- (void)testGroupIdentify {
    [self initAmpliWithNewInstance:@"testGroupIdentify"];

    NSString *groupType = @"test-group-type";
    NSString *groupName = @"test-group";
    
    [_ampli.client add:[AMPPlugin initWithType:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
        XCTAssertEqualObjects(event.eventType, @"$groupidentify");
        AMPProperties* groups = event.groups;
        XCTAssertEqual([groups get:groupType], groupName);
        NSMutableDictionary* groupPropertiesSet = [event.groupProperties get:@"$set"];
        XCTAssertEqual(groupPropertiesSet[@"requiredBoolean"], @false);
        XCTAssertEqual(groupPropertiesSet[@"optionalString"], @"optional string");
        [self->_pluginRun fulfill];
        return event;
    }]];

    AMPIdentify* identify = [AMPIdentify new];
    [identify set:@"requiredBoolean" value:@false];
    [identify set:@"optionalString" value:@"optional string"];
    [_ampli.client groupIdentify:groupType groupName:groupName identify:identify];
    [_ampli flush];

    [self waitForExpectationsWithTimeout:2.0 handler:^(NSError *error) {
        if (error) {
            NSLog(@"timeout errored: %@", error);
            XCTFail(@"test timed out");
        }
    }];
}

@end

File: ios/swift/v2/AmpliSwiftSampleApp/AmpliSwiftSampleAppTests/AmpliTests.swift
import XCTest
import AmpliSwiftSampleApp
import AmplitudeSwift

let emptyDictionary = [String: String]()
func isEmptyDictionary(_ dict: Any?) -> Bool {
    dict == nil || NSDictionary(dictionary: dict as! [String: String]).isEqual(to: emptyDictionary)
}

class AmpliTests: XCTestCase {
    private var ampli: Ampli!

    override func setUpWithError() throws {
        ampli = Ampli()
    }
    
    func initAmpliWithNewInstance(_ instanceName: String) {
        ampli.load(LoadOptions(client: LoadClientOptions(config: Configuration(
            apiKey: "test-api-key",
            instanceName: instanceName,
            trackingSessionEvents: false,
            migrateLegacyData: false
        ))))
    }

    func testIdentify() throws {
        let userId = "test-user-id";
        let deviceId = "test-device-id";
        
        initAmpliWithNewInstance("testIdentify")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        let identifyProperties = Identify(requiredNumber: 22.0, optionalArray: ["optional array str"])
        let eventOptions = EventOptions(deviceId: deviceId)

        ampli.identify(userId, identifyProperties, options: eventOptions)
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "$identify")
        XCTAssertTrue(isEmptyDictionary(event.eventProperties))
        XCTAssertEqual(event.userId, userId)
        XCTAssertEqual(event.deviceId, deviceId)
    }
    
    func testIdentifyUserIdOnEvent() {
        initAmpliWithNewInstance("testIdentifyUserIdOnEvent")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        let eventOptionsUserId = "test-user-id-options"
        let deviceId = "test-device-id"
        let identifyProperties = Identify(requiredNumber: 22.0, optionalArray: ["optional array str"])
        let eventOptions = EventOptions(userId: eventOptionsUserId, deviceId: deviceId)

        ampli.identify(nil, identifyProperties, options: eventOptions)
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "$identify")
        XCTAssertEqual(event.userId, eventOptionsUserId)
        XCTAssertEqual(event.deviceId, deviceId)
        XCTAssertEqual(ampli.client.getUserId(), eventOptionsUserId)
    }

    func testTrackWithNoProperties() throws {
        initAmpliWithNewInstance("testTrackWithNoProperties")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        ampli.eventNoProperties()
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "Event No Properties")
        XCTAssertTrue(isEmptyDictionary(event.eventProperties))
        XCTAssertNil(event.userId)
        XCTAssertNotNil(event.deviceId)
    }

    func testTrackEventWithAllTypes() throws {
        let userId = "test-user-id";
        let deviceId = "test-device-id";
        let eventOptions = EventOptions(deviceId: deviceId)

        initAmpliWithNewInstance("testTrackEventWithAllTypes")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        ampli.track(
            EventWithAllProperties(
                requiredArray: ["array element 1", "array element 2"],
                requiredBoolean: true,
                requiredEnum: EventWithAllProperties.RequiredEnum.enum1,
                requiredInteger: 10,
                requiredNumber: 2.0,
                requiredString: "required string"
            ).options(userId: userId),
            options: eventOptions
        )
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "Event With All Properties")
        let eventProperties = event.eventProperties
        XCTAssertNotNil(event.eventProperties)
        XCTAssertEqual(eventProperties!["requiredArray"] as! Array, ["array element 1", "array element 2"])
        XCTAssertEqual(eventProperties!["requiredBoolean"] as! Bool, true)
        XCTAssertEqual(eventProperties!["requiredEnum"] as! String, "Enum1")
        XCTAssertEqual(eventProperties!["requiredInteger"] as! Int, 10)
        XCTAssertEqual(eventProperties!["requiredNumber"] as! Double, 2.0)
        XCTAssertEqual(eventProperties!["requiredString"] as! String, "required string")
        XCTAssertFalse(eventProperties!.keys.contains("optionalString"))
        XCTAssertEqual(event.userId, userId)
        XCTAssertEqual(event.deviceId, deviceId)
    }

    func testSetGroup() throws {
        let groupType = "test-group-type";
        let groupName = "test-group";

        initAmpliWithNewInstance("testSetGroup")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        ampli.client.setGroup(groupType: groupType, groupName: groupName)
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "$identify")
        XCTAssertTrue(isEmptyDictionary(event.eventProperties))
         let userProperties = event.userProperties
         let userPropertiesSet = userProperties!["$set"] as? Dictionary<String, Any>
         XCTAssertEqual(userPropertiesSet![groupType] as! String, groupName)
    }

    func testGroupIdentify() throws {
        let groupType = "test-group-type";
        let groupName = "test-group";

        initAmpliWithNewInstance("testGroupIdentify")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        let groupProperties: [String:Any] = [
            "requiredBoolean": false,
            "optionalString": "optional str"
        ]
        ampli.client.groupIdentify(groupType: groupType, groupName: groupName, groupProperties: groupProperties)
        ampli.flush()

        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "$groupidentify")
        XCTAssertTrue(isEmptyDictionary(event.eventProperties))
        XCTAssertTrue(isEmptyDictionary(event.userProperties))
        let groups = event.groups
        XCTAssertEqual(groups![groupType] as! String, groupName)
        let groupPropertiesSet = event.groupProperties!["$set"] as? Dictionary<String, Any>
        XCTAssertEqual(groupPropertiesSet!["requiredBoolean"] as! Bool, false)
        XCTAssertEqual(groupPropertiesSet!["optionalString"] as! String, "optional str")
    }

    func testGroupIdentifyNilOptionalString() throws {
        let groupType = "test-group-type";
        let groupName = "test-group";

        initAmpliWithNewInstance("testGroupIdentifyNilOptionalString")
        let eventCollector = EventCollectorPlugin()
        ampli.client.add(plugin: eventCollector)

        let identify = Identify()
        identify.set(property: "requiredBoolean", value: false)
        identify.set(property: "optionalString", value: nil)
        ampli.client.groupIdentify(groupType: groupType, groupName: groupName, identify: identify)
        ampli.flush()


        XCTAssertEqual(eventCollector.events.count, 1)
        let event = eventCollector.events[0]
        XCTAssertEqual(event.eventType, "$groupidentify")
        XCTAssertTrue(isEmptyDictionary(event.eventProperties))
        XCTAssertTrue(isEmptyDictionary(event.userProperties))
        let groups = event.groups
        XCTAssertEqual(groups![groupType] as! String, groupName)
        let groupPropertiesSet = event.groupProperties!["$set"] as? Dictionary<String, Any>
        XCTAssertEqual(groupPropertiesSet!["requiredBoolean"] as! Bool, false)
        XCTAssertNil(groupPropertiesSet!["optionalString"])
    }
}

class EventCollectorPlugin: Plugin {
    var type: PluginType
    var amplitude: Amplitude?
    var events: [BaseEvent] = Array()

    init() {
        self.type = .destination
    }

    func setup(amplitude: Amplitude) {
        self.amplitude = amplitude
    }

    func execute(event: BaseEvent) -> BaseEvent? {
        events.append(event)
        return event
    }
}
